### ISSUE TITLE
Decide whether we should use release serving or nightly serving for knative-gcp master e2e tests

### ISSUE THREAD

**Issue by capri-xiyue** [google/knative-gcp#1534]
**Problem**
In https://github.com/google/knative-gcp/pull/1533, we changed from nightly serving to release serving to unblock the broken CI/CD.  Need to discuss whether we should use release serving or nightly serving for knative-gcp master e2e tests
(presubmit jobs)
**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Which persona is this feature for?

**Exit Criteria**
Figure out whether we should use release serving or nightly serving for knative-gcp master e2e tests,
if we decide to use release serving, we should figure out what goes wrong with nightly serving and try to switch back from release serving to nightly serving

**Time Estimate (optional):** 
How many developer-days do you think this may take to resolve?

**Additional context (optional)**
Add any other context about the feature request here.

**Comment by capri-xiyue** [comment]
fyi @liu-cong @chizhg

**Comment by liu-cong** [comment]
My Thoughts: We should at least try not to use nightly builds in our presubmit jobs. An unstable nightly can block all our PRs.

### Option 1:
* Use latest version of release (we need a helper function to get that)
* If we think it's important to make sure latest knative-gcp works with serving, then we can have a separate test
### Option 2:
* Use nightly in nightly jobs
* Use latest release in all other jobs

**Comment by chaodaiG** [comment]
Thinking this from a slightly different point of view, how about treating serving and eventing as third party dependencies?

From this angle, we assume that we don't have control over them, so a reasonable approach imo is that knative-gcp only need to maintain compatibility with LTS of serving and eventing, and if knative-gcp suddenly isn't compatible with them it needs to be fixed asap, and it worths blocking presubmit tests

TL;DR; is only relying on released serving and eventing. Note that it will cause delays in cutting releases

**Comment by liu-cong** [comment]
I think it might be challenging to treat eventing as a third party dependency given the current pace of development. But serving should have no problem.

**Comment by chizhg** [comment]
Sorry what do you guys mean by "third party dependencies" here? Copying the latest release yaml files to this repo and installing from there?

For the ways forward, I'm more leaning towards the option 1 proposed by Cong above:
- Change the presubmit test jobs to use the latest serving release, e.g. the current latest serving release is v0.16.0
- Add an optional presubmit job that uses the latest nightly serving release, so it won't block presubmit if it fails, but will provide signals that some compatibility issues need to be fixed before the final release.

**Comment by chaodaiG** [comment]
> Sorry what do you guys mean by "third party dependencies" here? Copying the latest release yaml files to this repo and installing from there?

Treating them something like GKE, i.e. the "latest" GKE being tested against is actually LTS of GKE, which is released from k8s release branches

**Comment by capri-xiyue** [comment]
If we don't use latest serving for our presubmit jobs, I'm not sure whether it will delay all the work related to compatibility issues to the final release cut.

Previously, when we find the serving or eventing has some compatibility issues with latest knative-gcp, to unblock knative-gcp CI/CD, we will ping serving and eventing to latest working nightly version to unblock CI/CD and then try to solve the compatibility issues ASAP.

I agreed with what @grant mentioned offline, "testing on nightly hides compatibility issues with the latest release. I guess we should really be testing both."
I think maybe we should add both tests which tests with nightly serving and eventing and also tests with latest release serving and eventing.

**Comment by liu-cong** [comment]
How do we depend on serving currently? Is serving only needed in the tests?

**Comment by chaodaiG** [comment]
I can totally see the point of delaying in cutting releases, which is a very solid issue. But still, I'm not feeling great in thinking of blocking contributions based on moving targets. So, how about:

- Gate presubmit test with LTS serving and eventing
- Continuously testing master with nightly serving and eventing
- (Optionally) When in doubt, one can manually trigger tests against nightly serving and eventing in their PR(s)

**Comment by capri-xiyue** [comment]
> How do we depend on serving currently? Is serving only needed in the tests?

I just checked the code. Besides the e2e tests, we also need ksvc when we create topic:
https://github.com/google/knative-gcp/blob/7bf16c6b11e72c295118cf8017e6fff882134b7a/pkg/reconciler/intevents/topic/topic.go#L212

**Comment by liu-cong** [comment]
If our code depend on certain serving release, I think the test clusters should install that release too. Right now we use 0.16.1: https://github.com/google/knative-gcp/blob/master/go.mod#L33

**Comment by capri-xiyue** [comment]
> If our code depend on certain serving release, I think the test clusters should install that release too. Right now we use 0.16.1: https://github.com/google/knative-gcp/blob/master/go.mod#L33

Yep. We are using https://github.com/google/knative-gcp/blob/f9e4a707ba1939c194f94a5e2b7c3a3cd02bc758/go.mod#L34.
It looks like it's most recent one. It has 20200810114129. Looks like the serving dependency is master which matches the nightly release.

**Comment by nlopezgi** [comment]
Hi folks, are there any updates on this issue? Is this still a problem?

**Comment by liu-cong** [comment]
I think we encountered another serving nightly issue since this issue was created. @capri-xiyue can confirm I think.

In my opinion it's pretty bad to block PRs due to nightly serving errors. But we may be OK with it given its relatively low frequency.

**Comment by capri-xiyue** [comment]
> I think we encountered another serving nightly issue since this issue was created. @capri-xiyue can confirm I think.
> 
> In my opinion it's pretty bad to block PRs due to nightly serving errors. But we may be OK with it given its relatively low frequency.

Yep, we encountered serving issues since istio had some problems. But the frequency is quite low, I think it will be ok if we use nightly serving.

**Comment by nlopezgi** [comment]
sgtm, closing this issue for now. Please reopen anytime if the frequency of errors increases and we need to have this discussion again. For what its worth, I agree with the principle that we should be using nightly serving to try to catch any issues as soon as they occur.
